File: client/src/types/messages.ts
export interface User {
  id: string;
  name: string;
}

export interface InitStateMessage {
  type: 'init_state';
  xml: string;
  revision: number;
  users: User[];
  locks: Record<string, string>;
  current_user_id: string;
}

export interface UsersMessage {
  type: 'users';
  users: User[];
}

export interface BpmnUpdateMessage {
  type: 'bpmn_update';
  xml: string;
  revision: number;
  from: string;
}

export interface LockMessage {
  type: 'lock';
  elementId: string;
  userId: string;
}

export interface UnlockMessage {
  type: 'unlock';
  elementId: string;
  userId: string;
}

export interface LocksSyncMessage {
  type: 'locks';
  locks: Record<string, string>;
}

export type IncomingMessage =
  | InitStateMessage
  | UsersMessage
  | BpmnUpdateMessage
  | LockMessage
  | UnlockMessage
  | LocksSyncMessage;

export type OutgoingMessage =
  | {
      type: 'bpmn_update';
      xml: string;
    }
  | {
      type: 'lock';
      elementId: string;
    }
  | {
      type: 'unlock';
      elementId: string;
    };

File: server/schema/messages.py
from typing import Dict, List, Optional
from pydantic import BaseModel, Field, ConfigDict
from typing import Literal

"""
Pydantic models describing the WebSocket message protocol used by the
ConnectionManager. These schemas are primarily used in tests and debugging
to validate the structure of messages emitted by the backend during
real-time collaboration.
"""

class User(BaseModel):
    id: str
    name: str


class BaseMessage(BaseModel):
    type: str

    model_config = ConfigDict(
        extra="ignore",
        populate_by_name=True
    )


class InitStateMessage(BaseMessage):
    type: str = "init_state"
    xml: str
    revision: int
    users: List[User]
    locks: Dict[str, str]


class UsersMessage(BaseMessage):
    type: str = "users"
    users: List[User]


class BpmnUpdateMessage(BaseMessage):
    type: str = "bpmn_update"
    xml: str
    revision: int
    from_user: Optional[str] = Field(default=None, alias="from")


class LockMessage(BaseMessage):
    type: str = "lock"
    elementId: str
    userId: str


class UnlockMessage(BaseMessage):
    type: str = "unlock"
    elementId: str
    userId: str

class LocksSyncMessage(BaseModel):
    type: Literal["locks"]
    locks: Dict[str, str]
